### PR TITLE
Laboratory Work #1, Irina Kurnakova - 20FPL1

### PR DESCRIPTION
**Pull request by IrinaKurnakova**

### CODE REVIEW

File: page_code.html [hunk 1]
```diff
@@ -0,0 +1,504 @@
+<!DOCTYPE html>
```
Comment (marina-kaz) [inline page_code.html:1]: we do not need html files to be saved as the goal of this lab is to collect a dataset that consists of texts

File: pipeline.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 Pipeline for text processing implementation
 """
 
+import requests
```
Comment (marina-kaz) [inline pipeline.py:5]: It is highly unlikely that you are going to need requests to process texts :)

File: scrapper.py [hunk 3]
```diff
@@ -26,16 +36,40 @@ class Crawler:
...
+        content = article_bs.find_all('div', id='main-content')
+        for article in content:
+            all_links = article.find_all('a')
+
+        for link in all_links:
+            try:
+                if link['href'][:4] == 'http':
+                    self.urls.append(link['href'])
+                else:
+                    self.urls.append('https://vja.ruslang.ru' + link['href'])
```
Comment (marina-kaz) [inline scrapper.py:53]: it is better to use this method to solely extract urls from given soup
leave adding it to the attribute to find_articles
also, before adding a new url to the attribute it is necessary to check that such url has not been added before as many sources duplicate urls on their pages

File: scrapper.py [hunk 4]
```diff
@@ -44,18 +78,58 @@ def get_search_urls(self):
...
+    def _fill_article_with_text(self, article_bs):
+        pass
+
+
 def prepare_environment(base_path):
     """
     Creates ASSETS_PATH folder if not created and removes existing folder
     """
-    pass
+    try:
```
Comment (marina-kaz) [inline scrapper.py:101]: this will work
however, is there a way to implement this without try/except tricks?

File: scrapper.py [hunk 5]
```diff
@@ -44,18 +78,58 @@ def get_search_urls(self):
...
+    with open(crawler_path) as file:
+        config = json.load(file)
+
+    seed_urls = config['seed_urls']
+    total_articles = config['total_articles_to_find_and_parse']
+
+    if not isinstance(total_articles, int) or total_articles <= 0:
+        raise IncorrectNumberOfArticlesError
+
+    if total_articles > 100:
```
Comment (marina-kaz) [inline scrapper.py:120]: I recommend taking a bigger number as 100 is a required number of articles to find and parse, so it is a bit odd that this code is expected to work for 100 items but is going to complain about 101

**Review by marina-kaz — CHANGES_REQUESTED** [review]
There is not much time left so I advise to speed up a bit :)

File: requirements.txt [hunk 6]
```diff
@@ -1 +1,5 @@
-
+beautifulsoup4==4.10.0
+pymupdf
+pytest
+requests==2.27.1
+wget
```
Comment (marina-kaz) [inline requirements.txt:5]: is there a reason why you do not specify version for the libraries?

File: scrapper.py [hunk 7]
```diff
@@ -2,6 +2,20 @@
 Scrapper implementation
 """
 
+import datetime
+import json
+import random
+import re
+import shutil
+import time
```
Comment (marina-kaz) [inline scrapper.py:10]: it is better to sort imports alphabetically

File: scrapper.py [hunk 8]
```diff
@@ -26,38 +40,124 @@ class Crawler:
...
+        article_author = article_bs.find('strong').text
+        self.article.author = article_author
+
+        meta_inf = re.findall(r'\d.+', article_bs.find('link', rel='alternate')['href'])[0]
+        year = meta_inf[:4]
+        volume = meta_inf[5]
+        months_dict = {'1': '01', '2': '03', '3': '05', '4': '07', '5': '09', '6': '11'}
+        month = months_dict[volume]
+        date_raw = f'{year}-{month}'
+        article_date = datetime.datetime.strptime(date_raw, '%Y-%m')
```
Comment (marina-kaz) [inline scrapper.py:112]: there is a method for that in article.py

File: scrapper.py [hunk 9]
```diff
@@ -26,38 +40,124 @@ class Crawler:
...
     Validates given config
     """
-    pass
+    with open(crawler_path) as file:
+        config = json.load(file)
+
+    seed_urls = config['seed_urls']
+    total_articles = config['total_articles_to_find_and_parse']
+
+    if not isinstance(total_articles, int) or total_articles <= 0:
```
Comment (marina-kaz) [inline scrapper.py:135]: separate these checks: if total_articles is not an integer this line will fail

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Good job, keep going

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Address the comments and you will be all set

File: scrapper.py [hunk 10]
```diff
@@ -26,38 +40,124 @@ class Crawler:
...
+        raise IncorrectNumberOfArticlesError
+
+    if total_articles > 200:
+        raise NumberOfArticlesOutOfRangeError
+
+    pattern = re.compile(r"^https?://")
+
+    for url in seed_urls:
```
[the remark this answers is not in the document]
Comment (IrinaKurnakova) [reply scrapper.py:155]: Done

File: scrapper.py [hunk 11]
```diff
@@ -26,38 +40,124 @@ class Crawler:
...
+        if not re.match(pattern, url):
+            raise IncorrectURLError
+
+    if not seed_urls:
+        raise IncorrectURLError
+
+    prepare_environment(ASSETS_PATH)
```
[the remark this answers is not in the document]
Comment (IrinaKurnakova) [reply scrapper.py]: Done

File: page_code.html (hunk 1 above)
Comment (marina-kaz) [reply page_code.html:1]: are you planning on deleting it? :)

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Congrats on successful Crawler submission!
Time to embark on pipeline

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to work on Pipeline :)

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Time to implement Pipeline

File: pipeline.py [hunk 12]
```diff
@@ -24,25 +30,28 @@ class MorphologicalToken:
...
         """
-        pass
+        return f'{self.normalized_form}<{self.tags_mystem}>'
 
     def get_multiple_tagged(self):
         """
         Returns normalized lemma with PyMorphy tags
         """
-        pass
+        return f'{self.normalized_form}<{self.tags_pymorphy}>'
```
Comment (marina-kaz) [inline pipeline.py:54]: according to [sample file](https://github.com/fipl-hse/2021-2-level-ctlr/blob/main/config/test_files/reference_score_eight_test.txt), multiple-tagged sequences are expected to contain both mystem and pymorphy annotations

File: pipeline.py [hunk 13]
```diff
@@ -72,26 +88,76 @@ class TextProcessingPipeline:
...
+        raise NotADirectoryError
+
+    if len(list(path_to_validate.glob('*'))) == 0:
+        raise EmptyDirectoryError
+
+    counter_txt = 0
+    counter_meta = 0
+
+    for file in path_to_validate.glob('*'):
+        if str(file)[-7:] == 'raw.txt':
```
Comment (marina-kaz) [inline pipeline.py:148]: here and in other places, consider using `endswith` [method](https://docs-python.ru/tutorial/operatsii-tekstovymi-strokami-str-python/metod-str-endswith/): it is designed exactly for this purpose and your code will be more readable

File: pipeline.py [hunk 14]
```diff
@@ -51,19 +60,26 @@ class CorpusManager:
...
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        path_to_raw_txt_data = Path(self.path_to_raw_txt_data)
+
+        for file in path_to_raw_txt_data.glob('*'):
+            if '_raw.txt' in str(file):
```
Comment (marina-kaz) [inline pipeline.py:74]: consider using `file.name`, it is designed especially for such cases and is more readable

File: pipeline.py [hunk 15]
```diff
@@ -72,26 +88,76 @@ class TextProcessingPipeline:
...
+            with open(file, 'r', encoding='utf-8') as current_file:
+                text = current_file.read()
+            if not text:
+                raise InconsistentDatasetError
+
+        if str(file)[-9:] == 'meta.json':
+            counter_meta += 1
+
+    if counter_txt != counter_meta:
+        raise InconsistentDatasetError
```
Comment (marina-kaz) [inline pipeline.py:161]: you should also ensure that files have correct numeration from 1 to N without any slips

**Review by marina-kaz — CHANGES_REQUESTED** [review]
You are off to a great start! 
Keep going

File: pipeline.py (hunk 15 above)
Comment (IrinaKurnakova) [reply pipeline.py:186]: Done

File: pipeline.py [hunk 16]
```diff
@@ -51,19 +62,26 @@ class CorpusManager:
...
+        self._scan_dataset()
 
     def _scan_dataset(self):
         """
         Register each dataset entry
         """
-        pass
+        path_to_raw_txt_data = Path(self.path_to_raw_txt_data)
+
+        for file in path_to_raw_txt_data.glob('*'):
```
Comment (marina-kaz) [inline pipeline.py]: there is a `Path` method for iterating over all the files in a directory
`.glob()` is more suitable for the cases where you need to filter directory content

**Review by marina-kaz — CHANGES_REQUESTED** [review]
Great job

**Pull request closed without merging**